### PR TITLE
Carousel: use buttons, not links, for prev/next controls

### PR DESCRIPTION
**Pull request by patrickhlauke**
- expand the styles to neutralise border/background
- change docs page
- add extra unit test to check that links or buttons work as controls
- modify visual test to use buttons as well

Preview https://deploy-preview-32627--twbs-bootstrap.netlify.app/docs/5.0/components/carousel/

### CODE REVIEW

**Comment by patrickhlauke** [comment]
to clarify: using links still works, but it's not really semantically correct. not a breaking change, as the old markup still works. just better/cleaner to do it using `<button>`

**Comment by patrickhlauke** [comment]
this will need to be expanded if https://github.com/twbs/bootstrap/pull/32638 lands first

**Comment by patrickhlauke** [comment]
temporarily set to draft while i expand this following merging of  #32638

**Review by mdo — APPROVED** [review]

**Comment by patrickhlauke** [comment]
setting to draft to make sure the added carousel section (disabled touch) is added appropriately

**Comment by patrickhlauke** [comment]
ah, my mistake, i see this won't be affected by my planned upcoming change. as you were

**Comment by XhmikosR** [comment]
There are more instances in examples?

Also, perhaps we should have callout/example mentioning that both work but for whatever reasons we suggest buttons?

File: site/content/docs/5.0/examples/carousel-rtl/index.html
Comment (XhmikosR) [inline site/content/docs/5.0/examples/carousel-rtl/index.html:62]: isn't the role needed since we use `a`?

**Comment by patrickhlauke** [comment]
> There are more instances in examples?

Ah, good catch, those slipped me by. Addressed now.

> Also, perhaps we should have callout/example mentioning that both work but for whatever reasons we suggest buttons?

the "for whatever reason" part of your comment worries me a bit. i thought we (at least in core team) all understood/aknowledged that links are for linking (to another page, or a section within the current page like a skip anchor) and buttons are for dynamic in-page actions. 

i'll add a little mention/note about this though in a moment, sure.

File: site/content/docs/5.0/examples/carousel-rtl/index.html
Comment (patrickhlauke) [reply site/content/docs/5.0/examples/carousel-rtl/index.html:62]: they're not buttons though, are they? they're links going to some page, no?

Comment (XhmikosR) [reply site/content/docs/5.0/examples/carousel-rtl/index.html:62]: Ah, true, those are captions. Might be worth splitting these to a separate PR?

Comment (patrickhlauke) [reply site/content/docs/5.0/examples/carousel-rtl/index.html:62]: to clarify: if a link acts as a link (goes to another page or whatever), rather than doing something dynamic (opening a modal dialog, a dropdown, etc), then it should *not* have `role="button"`. and for these examples, I'm assuming these links are meant to signify links that act as links. whether the link is visually styled as a `btn` is usually irrelevant ("call-to-action links"), what counts is what the purpose of the control is.

**Comment by patrickhlauke** [comment]
> i'll add a little mention/note about this though in a moment, sure.

done

File: site/content/docs/5.0/examples/carousel-rtl/index.html
Comment (patrickhlauke) [reply site/content/docs/5.0/examples/carousel-rtl/index.html:62]: > Might be worth splitting these to a separate PR?

admittedly, i'm already getting lost with the various PRs i had going at the moment, and as this is just tweaking something in an example, in passing, not sure if it's really useful to split? your call though.

**Comment by XhmikosR** [comment]
> the "for whatever reason" part of your comment worries me a bit. i thought we (at least in core team) all understood/aknowledged that links are for linking (to another page, or a section within the current page like a skip anchor) and buttons are for dynamic in-page actions.

I'm with you, but people might wonder, that's all I meant :)

File: site/content/docs/5.0/examples/carousel-rtl/index.html
Comment (patrickhlauke) [reply site/content/docs/5.0/examples/carousel-rtl/index.html]: ok reverted this change with the latest commit

Comment (XhmikosR) [reply site/content/docs/5.0/examples/carousel-rtl/index.html]: Thanks! Don't forget to submit a separate PR please

Comment (patrickhlauke) [reply site/content/docs/5.0/examples/carousel-rtl/index.html]: done https://github.com/twbs/bootstrap/pull/32788

**Review by ffoodd — APPROVED** [review]

File: site/content/docs/5.0/components/carousel.md
Comment (mdo) [inline site/content/docs/5.0/components/carousel.md:49]: Can make this short and sweet IMO.

```suggestion
Adding in the previous and next controls. We recommend using `<button>` elements, but you can also use `<a>` elements with `role="button"`.
```

**Review by mdo — APPROVED** [review]
One suggested edit

File: site/content/docs/5.0/components/carousel.md
Comment (patrickhlauke) [reply site/content/docs/5.0/components/carousel.md:49]: that might be misleading, as it may suggest our code only supports `<a>` with `role="button"`, while in effect it doesn't really check for `role="button"`?

Comment (patrickhlauke) [reply site/content/docs/5.0/components/carousel.md:49]: @mdo would you be ok with the more verbose, but possibly clearer, wording?

Comment (mdo) [reply site/content/docs/5.0/components/carousel.md:49]: I think brevity and best practice should win out—if folks use anchor elements, they really should use `role="button"`, yes? If that's not the case, then yeah we can use the more verbose.

Comment (patrickhlauke) [reply site/content/docs/5.0/components/carousel.md]: *shrug* fairy snuff :)

**Comment by XhmikosR** [comment]
@patrickhlauke Is it intentional that the carousel controls sort of jump when clicked? Firefox 85 on Windows 10. Edge Chromium seems to be the same as before.

**Comment by patrickhlauke** [comment]
> @patrickhlauke Is it intentional that the carousel controls sort of ~jump~move when clicked? Firefox 85 on Windows 10. Edge Chromium seems to be the same as before.

ah, good catch, hadn't even checked. no, wasn't intentional. forgot about Firefox's quirky behaviour of adding extra padding on `:active`. neutralised this now by explicitly setting `padding: 0` for the prev/next controls

File: js/tests/unit/carousel.spec.js
Comment (XhmikosR) [inline js/tests/unit/carousel.spec.js:1189]: @patrickhlauke just to confirm, so in this branch we no longer have a JS test for `<div class="carousel-control-prev"` etc?

**Comment by XhmikosR** [comment]
Now, it looks good to me, thanks! One last comment and also I wonder, do we have an issue about this?

File: js/tests/unit/carousel.spec.js
Comment (patrickhlauke) [reply js/tests/unit/carousel.spec.js:1189]: correct. happy to add another test back in that tests explicitly for `<div role="button" tabindex="0" ...>` but that shouldn't really behave any different

**Comment by patrickhlauke** [comment]
> Now, it looks good to me, thanks! One last comment and also I wonder, do we have an issue about this?

we have this meta-issue where i reel off various things that should get fixed https://github.com/twbs/bootstrap/issues/31186 but no atomic issue about just this aspect i think. can do a trawl through open issues after this just to check though

File: js/tests/unit/carousel.spec.js
Comment (XhmikosR) [reply js/tests/unit/carousel.spec.js:1189]: Yeah, agreed. In JS we just look for the class.

**Comment by patrickhlauke** [comment]
:+1: thanks for sanity checking @XhmikosR

**Pull request merged**
